common/src/main/java/io/netty/util/NetUtil.java

```diff
@@ -679,6 +634,50 @@ private static boolean isValidNumericChar(char c) {
         return c >= '0' && c <= '9';
     }
 
+    private static boolean isValidIPv4MappedChar(char c) {
+        return c == 'f' || c == 'F';
+    }
+
+    private static boolean isValidIPv4MappedSeparators(byte b0, byte b1) {
+        // We allow IPv4 Mapped (https://tools.ietf.org/html/rfc4291#section-2.5.5.1)
+        // and IPv4 compatible (https://tools.ietf.org/html/rfc4291#section-2.5.5.1).
+        // The IPv4 compatible is deprecated, but it allows parsing of plain IPv4 addressed into IPv6-Mapped addresses.
+        return b0 == b1 && (b0 == 0 || b0 == -1);
+    }
+
+    private static boolean isValidIPv4Mapped(byte[] bytes, int currentIndex) {
+        if (currentIndex > 12 || currentIndex < 2 ||
+            !isValidIPv4MappedSeparators(bytes[currentIndex - 1], bytes[currentIndex - 2])) {
+            return false;
+        }
+        currentIndex -= 2;
```

Comment:
Maybe `-= 3`? With `-2` test failed for ip "::ffff:1.2.3.4" (whithout `Unsafe`).